internal/pkg/debug/graph/graph.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package graph
+
+import (
+	"golang.org/x/tools/go/ssa"
+)
+
+type NodeRelationship int
+
+const (
+	Referrer NodeRelationship = iota
+	Operand
+)
+
+type RelatedNode struct {
+	N ssa.Node
+	R NodeRelationship
+}
+
+type Graph struct {
+	// the function whose graph this is
+	F *ssa.Function
+	// a mapping from each node to its neighbors (referrers + operands)
+	Neighbors map[ssa.Node][]RelatedNode
+	// this is used while creating the graph to avoid needlessly revisiting nodes
+	visited map[ssa.Node]bool
+}
+
+// New returns a new Graph constructed from a given function.
+func New(f *ssa.Function) Graph {
+	g := Graph{
+		F:         f,
+		Neighbors: map[ssa.Node][]RelatedNode{},
+		visited:   map[ssa.Node]bool{},
+	}
+	g.visitBlocks()
+	return g
+}
+
+func (g *Graph) visitBlocks() {
+	for _, b := range g.F.Blocks {
+		g.visit(b)
+	}
+}
+
+func (g *Graph) visit(b *ssa.BasicBlock) {
+	n := b.Instrs[0].(ssa.Node)
+
+	g.visited[n] = true
+
+	stack := []ssa.Node{n}
```

Comment:
Could we hide these details of operating on a stack to a dedicated type?

Reply:
Do you mean introducing a `Stack` type with push and pop methods? I'm not sure I see a benefit to doing that. I've seen this pattern of using a simple `slice` as a stack several times and it feels quite natural to me.

Reply:
I gave it a shot and I think it is quite nice. WDYT?